### ISSUE TITLE
Regression? ProDOS 2.5.0a4 date-handling is broken

### ISSUE THREAD

**Issue by apple2geek** [ProDOS-8/ProDOS8-Testing#20]
## Description

Files with dates are showing bizarre values, including "n/a" for month, and years with semicolons and inverse characters etc. in at least ProTERM3 and Super Selector.
These exact same files, on the exact same drive, show correct dates in the exact same applications (PT3 and SS) when ProDOS8 2.5.0a3 is booted.

Curiously, BASIC.SYSTEM 1.7a9 (included with P8 2.5.0a4) shows these dates correctly using a "catalog" command, even when ProDOS8 _2.5.0a4_ is booted.

Screenshots showing the behavior are attached to comments below.

## Expected Behavior

Dates should be consistent and correct across all .SYSTEM programs.

## Actual Behavior

At least PT3.SYSTEM and SELECTOR.SYSTEM show bizarre, incorrect/nonsense values for file dates when listing directory contents.  Correct values for the dates are shown for those same files, using the same applications, when ProDOS 2.5.0a3 is booted.

## Steps to Reproduce
1. Boot ProDOS8 2.5.0a4
2. Launch a .system program (PT3.SYSTEM and SELECTOR.SYSTEM are the ones I've used)
3. Catalog a directory, such as the root directory of the boot drive, using the program's facility to do so
4. Observe scrambled/nonsense date values
5. Reboot system using ProDOS8 2.5.0a3
6. Launch the SAME .system program, from the same media
7. Catalog the same directory
8. Observe correct, sane, date values

## Context

Just routine use of the system after updating PRODOS and BASIC.SYSTEM

### Hardware:

Physical Apple //e
ZipChip8
Focus Hard Drive
RamWorks III

### ProDOS8 version:
2.5.0a4 vs 2.5.0a3

### Non-Operating System release software being used: 
ProTERM 3.1, Super Selector 3.2.2

**Comment by apple2geek** [comment]
These three images are with ProDOS8, 2.5.0a3 booted, and show correct and consistent dates and times. (Sorry so big - exported them full size)
The root directory of the drive is used in each.  One is Super Selector 3.2.2, one is ProTERM 3.1, and one is Basic.System 1.7d8.
![2 5 0a3-SS](https://user-images.githubusercontent.com/54514423/64216399-f0535780-ce75-11e9-84b5-b2b2df85f8f0.png)
![2 5 0a3-PT3](https://user-images.githubusercontent.com/54514423/64216398-f0535780-ce75-11e9-82ee-48178a0d6b04.png)
![2 5 0a3-BASIC](https://user-images.githubusercontent.com/54514423/64216397-efbac100-ce75-11e9-90fa-63234878aa80.png)

**Comment by apple2geek** [comment]
These three are with 2.5.0a4 booted.  EXACT same hard drive, EXACT same binaries of each application, EXACT same directory listed.  The only difference is the version of ProDOS booted and the version of BASIC.SYSTEM (1.7a9) used.  Note the crazy, scrambled date/time values.
BASIC.SYSTEM seems to at least show proper dates, but the times are all zeroed.
![2 5 0a4-SS](https://user-images.githubusercontent.com/54514423/64216584-b0d93b00-ce76-11e9-8b2b-fa5593fccd8f.png)
![2 5 0a4-PT3](https://user-images.githubusercontent.com/54514423/64216583-b0d93b00-ce76-11e9-97aa-14efe3b03b4f.png)
![2 5 0a4-BASIC](https://user-images.githubusercontent.com/54514423/64216582-b040a480-ce76-11e9-8b19-cf716409298c.png)

**Comment by JohnMBrooks** [comment]
It is actually a feature.  :)

2.5 uses an extended date format to fix the Y2028 and Y02040 problems with previous P8 dates. The extended dates have a simple 12-bit year: 0CE - 4095 CE instead of the complex legacy date system which will start failing in about 8 years.

2.5.0 will convert the day/month/year of pre-2.5 date formats into the extended format on-the-fly when directories are read by apps. Any newly created file or modified file will be assigned an extended-format date by P8 2.5.

Apps which display the date/time from the P8 clock driver or from directory entries will need a mod so they display extended dates correctly. The legacy ProDOS 7-bit, 2-digit dates are doomed.

**Comment by JohnMBrooks** [comment]
The plan is to modify Mr.Fixit so it will convert legacy file dates into extended dates.
[ProDOS-2.5.0a4-extended-date-format.txt](https://github.com/ProDOS-8/ProDOS8-Testing/files/3572150/ProDOS-2.5.0a4-extended-date-format.txt)

**Comment by apple2geek** [comment]
What changed between 2.5.0a3 and 2.5.0a4 that allowed a3 to continue to display legacy dates properly?  Did a3 not include the code to handle the new format yet?

**Comment by JohnMBrooks** [comment]
Yes, a3 had the old clock drivers. a4 has the no-slot clock driver and updated thunderclock & IIGS clock driver with extended date format

**Comment by JohnMBrooks** [comment]
I wish Apple hadn't created such limited dates with ProDOS in the 80s, but they did. It's a pain to fix/change, but better to do it now when there are a few years to make changes before all the dates fail.

**Comment by apple2geek** [comment]
Is there any risk of the contents of existing directories/files being changed in such a way that older versions of ProDOS will crash or improperly handle them if a piece of media is later accessed on a legacy P8 (2.4.x or even 2.0.3) system?

**Comment by JohnMBrooks** [comment]
Not that I'm aware of. The dates are pretty much cosmetic as far as apps are concerned.

**Comment by JohnMBrooks** [comment]
It would be useful to open a ticket with a list of ProDOS apps which display dates incorrectly so we can get them queued up for mods.

**Comment by JohnMBrooks** [comment]
It was pretty easy to mod BASIC.SYSTEM to support extended dates. It was simpler & less code than the legacy format with it's wacky 7-bit year.

**Comment by oliverschmidt** [comment]
> Apps which display the date/time from the P8 clock driver or from directory entries will need a mod so they display extended dates correctly.

@JohnMBrooks:

- Are you going to provide guidance how an app should best determine if the old or the new date format is present?
- I seem to remember that at some point you thought about a new opendir()/readdir()/closedir() style API that would allow ProDOS to hide the actual directory entry layout from apps. Is it correct to presume that you rejected that approach?

Not directly related:

- Are you going to provide an overview of the changes visible to / relevant for apps?
- Are you going to provide detailed information about the changes (e.g. mixed case file names) in 'ProDOS-2.5.0a4-extended-date-format.txt' style?

**Comment by JohnMBrooks** [comment]
Good questions.

Under ProDOS 2.5, all dates are in the new format. The clock drivers return new dates and when directories with old-date files are opened and read by an app, the in-memory version of the day/month/year for each old-date directory entry is converted to the new date at the same time that the filenames are converted to mixed-case.

A modified P8 2.5+ version of an app can be made which only handles new P8 2.5 features, or if the app expects to run on older versions of ProDOS also, it can check the ProDOS version byte to determine whether to display dates as new format or old format and whether new 2.5 features are present.

I would like to add a DirEntry() API to ProDOS, but it would take too much code space for the current 12KB LC fixed memory space. At the moment, I am proceeding with adding small features and fixes which fit in the existing 12KB memory space.

Yes, I plan to document all the P8 changes which affect apps.
So far nearly all apps work with 2.5, but some have cosmetic (date or inverse lowercase names) or don't take advantage of new P8 2.5 features.

P8 2.5 changes likely to prompt an app mod:
1) allow users to enter  "../" in pathnames
2) lowercase filenames (P8 auto-disables lowercase on II, II+)
3) extended date w/12-bit year,

P8 2.5 changes unlikely to prompt an app mod:
4) clock drivers now return seconds and milliseconds in $BF00 page
5) new P8 boot blocks (ideally mod those apps which can format a boot disk)
6) max of two interrupt handlers (2.4 & earlier allow 4 interrupt handlers)

Regarding the mixed-case filenames, P8 now follows Apple's GS/OS mixed-case filename spec, so the format has already been documented. GS Finder mixed-case names are compatible with P8 2.5 and vice versa.

There will be a new GS/OS PRO.FST (see 2.5.0a4 disk) which allows >51 files in the root directory, will write the new boot blocks to formatted disks, and will (to-do) convert both old and new dates to GS/OS date format.

**Comment by oliverschmidt** [comment]
@JohnMBrooks:

Thanks for the comprehensive information!

I see a single cc65 C Library for the Apple II supporting both P8 < 2.5 and P8 >= 2.5.

If I get you right then as soon as $BFFF >= $25 an app can presume to see

- lowercase filenames
- 12-bit years

for all volumes - no matter how they were formatted - when opening/reading a directory file. That would be great and simplify things a lot for apps :-)

Regarding the "../" I'm a little unsure. I don't have any code in the C Library that prohibits it. My current logic for "converting" file names provided by the user into files to be used with P8 is:

```
if it starts with '/'
  use it as-is
else
  if there's a prefix set
    prepend the prefix
  else
    get the volume name of the last accessed device and prepend it (of course with a '/').
```

As far as I know that's what BASIC.SYSTEM does and therefore likely matches expectations best. Do you see a conflict of that logic with the introduction of "../"?

BTW: The C Library will for sure make use of clock drivers returning seconds and milliseconds in $BF00 page. Do I get it right that as soon as $BFFF >= $25 an app can presume $BF8E/$BF8F to be valid - at least set to zero in case some clock driver should happen to not actually support higher resolution?

**Comment by JohnMBrooks** [comment]
> I see a single cc65 C Library for the Apple II supporting both P8 < 2.5 and P8 >= 2.5.
> If I get you right then as soon as $BFFF >= $25 an app can presume to see ...

Correct.

> That would be great and simplify things a lot for apps :-)

Yes, the extended date greatly reduced the year display code in BASIC.SYSTEM since the year can just be printed via standard 16-bit decimal print routine which tend to be fairly common in apps.

> Regarding the "../" I'm a little unsure. I don't have any code in the C Library that prohibits it. 

Good. Some apps enforce the ProDOS pathname rules of only-alpha 1st char. Disabling app 'input blocking' code should be a pretty easy mod to legacy apps.

> My current logic for "converting" file names provided by the user into files to be used with P8 is:

FYI: ProDOS already does all that work (ie, leading '/' means full-path, else it is partial and is appended to the current prefix or to the last-accessed device's volume name if the prefix is only '/'). P8 2.5 also converts lowercase names to uppercase and constructs the 16-bit mixed-case bit array that is stored along with the uppercase name in the file's directory entry.

> Do you see a conflict of that logic with the introduction of "../"?

No, but I think that logic is superfluous. ProDOS 2.5 expects apps to pass mixed-case user paths & filenames and P8 will 'do the right thing' or return an error if the path is invalid (has spaces, too long, etc).

> Do I get it right that as soon as $BFFF >= $25 an app can presume $BF8E/$BF8F to be valid

Correct, though a clock driver must be active (MachID $BF98 bit 0 == 1), and as you mentioned, not all clock HW & drivers will have sub-second times, so those drivers will likely leave $BF8E and $BF8F as-is instead of zeroing them out since the clock drivers are currently squeezed into 125 bytes.

**Comment by oliverschmidt** [comment]
> ProDOS already does all that work [...] is partial and is appended [...] to the last-accessed device's volume name if the prefix is only '/').

Hm, I until now I was absolutely sure that this was done by BASIC.SYSTEM, not by ProDOS...

> [...] as you mentioned, not all clock HW & drivers will have sub-second times, so those drivers will likely leave $BF8E and $BF8F as-is instead of zeroing them out since the clock drivers are currently squeezed into 125 bytes.

Have you considered to have ProDOS always zero those too locations before calling into the clock driver so that apps calling GET_TIME see a consistent behavior?

**Comment by oliverschmidt** [comment]
@JohnMBrooks:

I wrote:

>> [...] get the volume name of the last accessed device and prepend it (of course with a '/').

You wrote:

> ProDOS already does all that work (ie, leading '/' means full-path, else it is partial and is appended to the current prefix _**or to the last-accessed device's volume name if the prefix is only '/')**_.

I'm unable to verify what you wrote regarding the last-accessed device. The program

```
.code

    jsr     $BF00
    .byte   $C8
    .addr   open

    jsr     $FDDA
    jsr     $FC66

    jsr     $BF00
    .byte   $CC
    .addr   close

    jmp     $03D0

.data

open:
    .byte   $03
    .addr   name
    .addr   $0800
    .byte   $00

name:
    .byte   $06, "PRODOS"

close:
    .byte   $01
    .byte   $00
```

fails with `40` meaning "Invalid pathname syntax. The pathname contains illegal characters." when run after a `PREFIX /`. But it succeeds after a `PREFIX /VOLUME`.

**Comment by JohnMBrooks** [comment]
You are correct, P8's path handling code will append a partial pathname to the current prefix, but only if the current prefix is valid (ie, pfixptr at $BF9A != 0). If there is no valid prefix ("/"), then only full paths are allowed.

If BF9A == 0, then BASIC uses the most recently accessed device at $BF30, performs a P8 $C5 ONLINE call to find the device's volume name, and creates a full path using the volume name of that device followed by the partial pathname.

Ideally this logic would move into P8 since it seems like generic OS functionality. However, like most improvements to P8, code space is currently the limiting factor.

**Comment by oliverschmidt** [comment]
@JohnMBrooks:

* Thanks for acknowledging my statement above that the functionality in question resides in BASIC.SYSTEM, not ProDOS. So we agree that the logic in the cc65 C Library is not superfluous.

* It would be great if you could comment on my proposal to have ProDOS zero $BF8E and $BF8F before calling into the clock driver.

**Comment by JohnMBrooks** [comment]
> So we agree that the logic in the cc65 C Library is not superfluous.
Partially. The C library does not need to prepend the prefix when a valid prefix is set.

ie, instead of the current logic described as:

if it starts with '/'
  use it as-is
else
  if there's a prefix set
    prepend the prefix
  else
    get the volume name of the last accessed device and prepend it (of course with a '/').

cc65 lib can do this simplified logic:

if it starts with '/'
  use it as-is
else
  if there's a prefix set
    use it as-is
  else
    get the volume name of the last accessed device and prepend it (of course with a '/').

I will look into how many bytes it would take for P8 to automatically prepend the volume name when no prefix is set. If I can squeeze that into P8 2.5, then the C lib & BASIC logic can simplify to:

if it starts with '/'
  use it as-is
else
  if there's a prefix set
    use it as-is
  else
    use it as-is

**Comment by JohnMBrooks** [comment]
> Have you considered to have ProDOS always zero those too locations before calling into the clock driver so that apps calling GET_TIME see a consistent behavior?

It is not practical with P8's current memory layout. The clock driver is called from multiple locations in P8, so zeroing the time needs to be done inside the clock driver for those drivers which do not return seconds or milliseconds. The current clock drivers do not have available space in their 125 byte memory region to zero out unused fields.

Once P8 moves to a 'loadable driver' model, post-2.5, these improvements should become possible.

**Comment by oliverschmidt** [comment]
>> Have you considered to have ProDOS always zero those too locations before calling into the clock driver so that apps calling GET_TIME see a consistent behavior?

> It is not practical with P8's current memory layout. [...]

I see. So do you consider it "correct" behavior of an app to zero $BF8E and $BF8F before calling GET_TIME (of course only after checking $BFFF for ProDOS >= 2.5)?

**Comment by oliverschmidt** [comment]
> cc65 lib can do this simplified logic:
if it starts with '/'
use it as-is
else
if there's a prefix set
use it as-is
else
get the volume name of the last accessed device and prepend it (of course with a '/').

I just re-checked the code and that's exactly the logic it already follows today - sorry for the confusion caused!

> I will look into how many bytes it would take for P8 to automatically prepend the volume name when no prefix is set.

While that would of course be great in general, the cc65 C Library specifically wouldn't benefit as it needs to be runtime-compatible with Pre-2.5 ProDOS.

**Comment by JohnMBrooks** [comment]
> > > Have you considered to have ProDOS always zero those too locations before calling into the clock driver so that apps calling GET_TIME see a consistent behavior?
> 
> > It is not practical with P8's current memory layout. [...]
> 
> I see. So do you consider it "correct" behavior of an app to zero $BF8E and $BF8F before calling GET_TIME (of course only after checking $BFFF for ProDOS >= 2.5)?

Yes, it is correct to zero the date/time and/or to set it to a default (usually user-provided) value in case the clock driver is not present or does not set the field.

Many early ProDOS apps will prompt the user for the date and set it directly into the $BF90-$BF93 vars if the P8 MachID var indicates that no clock is present.

**Comment by oliverschmidt** [comment]
> Yes, it is correct to zero the date/time and/or to set it to a default (usually user-provided) value in case the clock driver is not present or does not set the field.

I see. Thanks.

> Many early ProDOS apps will prompt the user for the date and set it directly into the $BF90-$BF93 vars if the P8 MachID var indicates that no clock is present.

The Apple II variant of the cc65 C Library offers clock_settime() for this very purpose :-)

**Comment by bobbimanners** [comment]
Is the new date format documented somewhere?  I am playing around with some code to convert back and forwards for testing purposes (in Aztec C).  I guess the legacy format is that documented in the ProDOS Technical Reference.  However what happens to skirt the Y2K issue?  There is some history here that I a not really familiar with.

**Comment by JohnMBrooks** [comment]
See Bredon's discussion of ProDOS dates and Y2K patches

http://www.apple2.org.za/gswv/USA2WUG/Glen.Bredon.In.Memoriam/A2.Software/A2.Y2K.Patcher/Y2K.explanation.txt

ProDOS dates fail at:
1) 2000 (apps which use 0-99 as 1900-1999 and are not Y2K patched)

Y2K patches:
2) 2028 (apps which use the 7-bit year 0-127 as 1900-2027)
3) 2039 (apps which use year 40-99 as 1940-1999 and 0-39 as 2000-2039)

P8 2.5 extended date format: 12-bit year 0CE to 4095 CE
[ProDOS-2.5.0a4-extended-date-format.txt](https://github.com/ProDOS-8/ProDOS8-Testing/files/3603665/ProDOS-2.5.0a4-extended-date-format.txt)

**Comment by bobbimanners** [comment]
Thanks John.  If I understand correctly P8 2.5 will convert to new format on-the-fly, so a disk can have a mix of the two formats?

**Comment by JohnMBrooks** [comment]
Correct. P8 2.5 auto-converts old dates to new dates (in memory) as directories are opened and read, so apps can be converted to a single (new) date format.

I also hope to make a P8 2.4.3 for interim use which will auto-convert new dates to old dates for users who have a mix of old date and new date apps.

**Comment by bobbimanners** [comment]
I am writing some simple tools for the Aztec C environment to convert a file date from old->new or vice versa.  This will be useful for test and it may grow into a bigger utility as time goes on.

**Comment by bobbimanners** [comment]
I did a bit of experimenting this evening with Aztec C and ProDOS dates and times, and I thought I would share it here.  I think my conclusion is that the Aztec C `getfinfo()` call returns the raw information from the ProDOS directory bypassing the conversion you are doing in P8 2.5.0a4 somehow.  This points to a possible problem where some apps are getting converted date/time info (in the new format) and others are using an interface which lets them get at the unconverted values.

My code consists of two C files which compile and link to create an Aztec C program called `filedate`.  This program is pretty simple:

1. Parse the filename passed on the command line.
2. Call `getfinfo()` on the file passed on the command line and store the result.
3. Print out ProDOS version byte.
4. Print out the two words in hex for each of the creation and modification times.
5. Interpreting the two words in legacy ProDOS date/time format, print out date and time for both the creation and modification time.
6. Interpreting the two words in the new ProDOS 2.5.0 date/time format, print out date and time for both the creation and modification time.

I ran this program under both ProDOS 2.4.2 and 2.5.0 alpha 4, using the file `/H1/PRODOS` as a target.  Here are the results:

```
ProDOS Version 25

Creation:     d=$242d t=$0909
Modification: d=$2432 t=$0700
Legacy ProDOS 1.0 to 2.4.2 date and time:
Creation:     18-01-13 09:09
Modification: 18-01-18 07:00
ProDOS 2.5.0 date and time:
Creation:     1069-01-01 08:09
Modification: 1074-01-00 56:00

ProDOS Version 24

Creation:     d=$242d t=$0909
Modification: d=$2432 t=$0700
Legacy ProDOS 1.0 to 2.4.2 date and time:
Creation:     18-01-13 09:09
Modification: 18-01-18 07:00
ProDOS 2.5.0 date and time:
Creation:     1069-01-01 08:09
Modification: 1074-01-00 56:00
```

As you can see the results are identical on both 2.4.2 and 2.5.0 alpha 4.  This is surprising because other programs, such as the `ls` provided with Aztec C show the correct time and date for files under 2.4.2, but garbage under 2.5.0 alpha 4.

So it seems that whatever Aztec C does to get the two words, it ends up with the same raw hex value under 2.5.0 alpha 4 as in for 2.4.2., and this value is presumably the raw unconverted value of the timestamps on disk.

It seems that your conversion does not catch all ways of getting ahold of this information, so we will have a problem where it is never possible to get all applications to work properly.  Or am I missing something?

Here is my simple code - `p8.date.c`:

```
/* ProDOS Date and Time Handling */

#include <stdio.h>

/* int is assumed to be 16 bit in this code */
typedef unsigned int uint16;

/* Display date & time for ProDOS 1.0 to 2.4.2 */
printLegacyDateTime(d, t)
uint16 d;
uint16 t;
{
        printf("%02d-%02d-%02d",
               (d&0xfe00) >> 9, /* Year  */
               (d&0x01e0) >> 5, /* Month */
               d&0x001f);       /* Day   */

        printf(" %02d:%02d\n",
               (t&0x1f00) >> 8,  /* Hour   */
               t&0x003f);        /* Minute */
}

/* Display date & time for ProDOS 2.5.0 and later */
printNewDateTime(d, t)
uint16 d;
uint16 t;
{
        printf("%02d-%02d-%02d",
               d&0x0fff,                /* Year  */
               ((d&0xf000) >> 12) - 1,  /* Month */
               (t&0xf800) >> 11);       /* Day   */

    printf(" %02d:%02d\n",
           (t&0x07c0) >> 5,       /* Hour   */
           t&0x003f);             /* Minute */
}
```
And `filedate.c`:

```
/* Print ProDOS file dates */

#include <stdio.h>
#include <prodos.h>

extern printLegacyDateTime();
extern printNewDateTime();

int main(argc, argv)
int argc;
char *argv[];
{
    struct finfo fi;
        char *ver = 0xbfff;

    if (argc != 2) {
        puts("usage: filedate filename");
        return 1;
    }
    if (getfinfo(argv[1], &fi) == -1) {
        puts("getfinfo() error");
        return 1;
    }

        printf("ProDOS Version %2x\n\n", *ver);
    printf("Creation:     d=$%04x t=$%04x\n", fi.create_date, fi.create_time);
    printf("Modification: d=$%04x t=$%04x\n", fi.mod_date, fi.mod_time);

    puts("Legacy ProDOS 1.0 to 2.4.2 date and time:");
    printf("Creation:     ");
    printLegacyDateTime(fi.create_date, fi.create_time);
    printf("Modification: ");
    printLegacyDateTime(fi.mod_date, fi.mod_time);

    puts("ProDOS 2.5.0 date and time:");
    printf("Creation:     ");
    printNewDateTime(fi.create_date, fi.create_time);
    printf("Modification: ");
    printNewDateTime(fi.mod_date, fi.mod_time);

    return 0;
}
```

**Comment by JohnMBrooks** [comment]
Ah, I am not auto-converting dates returned by GetFileInfo. I will fix that in 2.5.0a5.

**Comment by JohnMBrooks** [comment]
I believe it is fixed in 2.5.0a5

https://drive.google.com/file/d/1n6ksz5DPddFRx3f0swJV2xcNwdKKIAL_

a5 also supports up to 8 drives per slot (no more wacky slot remapping) and up to 37 total drives can be mounted at once.

**Comment by bobbimanners** [comment]
Hi John - it seems that the Google Drive link for 2.5.0a5 doesn't work.
